Project/src/pages/SpeakingMode/SpeakingMode.ts

```diff
@@ -0,0 +1,189 @@
+            this.recognition.onnomatch = (event => {
+               this.chooseIncorrect();
+            });
+
+            //either too much noise or something
+            this.recognition.onerror = (event => {
+                
+                this.presentToast("Opps. We didn't hear you correctly, please tap the microphone and pronounce the word again :)")
+            });
+
+            //able to pick up what you are saying
+
+            this.recognition.onresult = event => {
+                this.zone.run(() => { // <== added
+                    var i = 0;
+                    if (event.results.length > 0) {                        
+                        for (i = 0; i < event.results.length; i++) {
+                            //only choose words that are higher than 0.9 confidence
+
+                            //First word is usually the correct word pronounced, but the speech recognition will not be confident sometimes
+                            if (this.currUnit.word === event.results[0][0].transcript.toLowerCase()) {
+                                this.isCorrect = true;
+                                break;
+                            }
+                            //only if confidence level is over 90% for remaining words then it will accept as a right answer
+                            if (event.results[i][0].confidence >= 0.9) {
+                                if (this.currUnit.word === event.results[i][0].transcript.toLowerCase()) {
+                                    this.isCorrect = true;
+                                    break;
+                                }
+                            }
+                        }
+
+                        if (this.isCorrect === true) {
+
+                            this.chooseCorrect();
+                            this.autoAdvance();
+                            return;
+                        }
+						//means out of all the options, non is correct
+                        this.chooseIncorrect();
+
+                        
+                    }
+                });
+            };
+        });
+    }
+
+    presentToast(message) {
+        let toast = this.toastCtrl.create({
+            message: message,
+            position: "bottom",
+            duration: 3000
+        });
+        toast.present();
+    }
+
+    initUnit = function () {
+        // Logic for setting up a new screenUnit
+        // Maybe add statisitics tracking here?
+        var path = this.plt.is('android') ? cordova.file.applicationDirectory + 'www/' : ''; //might be a hack...
+        this.currState = this.state.init; // Go to initial state
+        this.currUnit = this.screenUnits[this.currIndex]; // Set current screenUnit
+
+        let randomIndex = Math.floor(Math.random() * this.currUnit.audioPaths.length); // Pick audio clip to use
+        this.plt.ready().then((readySource) => { // Make sure the platform is ready before we try to use native components
```

Comment:
Have a look at ListeningMode.ts. In ListeningMode, I moved this this.plt.ready() check into the constructor. If we leave it in initUnit, we're doing a platform check every time we init a new screen unit, but we don't need to. You can do the same thing here as we do in ListeningMode: 

1. Do the this.plt.ready().... check in the constructor and set a variable "this.loaded = true" when platform is ready.
2. Setup SpeakingMode.html like ListeningMode.html where you have:

```
<ion-content *ngIf="!loaded">
  <div class="loader">Loading...</div>
</ion-content>
```

and then the second <ion-content> (the real content) would have `*ngIf="loaded"`

This way, we only check platform once, and we don't allow the user to interact with anything until platform is ready. Let me know if you need any help!